crates/storage-macros/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+/// * `#[skip]`       - Can be applied to fields that should not be serialised
+///                     or included in the Merkle hash calculation. These fields
+///                     will be completely ignored by the storage system, and
+///                     not even have getters and setters implemented.
+/// * `#[storage]`    - Indicates that the field is the storage element for the
+///                     struct. This is a mandatory field, and if it is missing,
+///                     there will be a panic during compilation. The name is
+///                     arbitrary, but the type has to be an [`Element`](calimero_storage::entities::Element).
+///
+/// Note that fields marked with `#[private]` or `#[skip]` must have [`Default`]
+/// implemented so that they can be initialised when deserialising.
+///
+/// TODO: The `#[collection]` attribute is not yet implemented, and the
+///       `#[private]` attribute is implemented with the same functionality as
+///       `#[skip]`, but in future these will be differentiated.
+///
+/// # Getters and setters
+///
+/// The macro will generate getter and setter methods for each field. These
+/// methods will allow the struct to control access to its fields, and ensure
+/// that any changes to the fields are reflected in the [`Element`](calimero_storage::entities::Element)'s
+/// state.
+///
+/// The getter methods will have the same name as the field, and the setter
+/// methods will be prefixed with `set_`. For example, given a field `name`, the
+/// getter method will be `name()`, and the setter method will be `set_name()`.
+///
+/// The setter methods will return a boolean indicating whether the update was
+/// carried out. Note that this is more an indication of change than it is of
+/// error — if the value is the same as the current value, the update will not
+/// be carried out, and the method will return `false`.
+///
+/// # Examples
+///
+/// ```
+/// use calimero_storage::entities::Element;
+/// use calimero_storage_macros::AtomicUnit;
+///
+/// #[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
+/// #[children(Paragraph)]
+/// struct Page {
+///     title: String,
+///     #[private]
+///     secret: String,
+///     #[storage]
+///     storage: Element,
+/// }
+///
+/// #[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
+/// struct Paragraph {
+///     text: String,
+///     #[storage]
+///     storage: Element,
+/// }
+/// ```
+///
+/// TODO: Once multiple child types are supported, this example will represent
+///       the correct approach.
+///
+/// ```ignore
+/// use calimero_storage::entities::Element;
+/// use calimero_storage_macros::AtomicUnit;
+///
+/// #[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
+/// struct Person {
+///     name: String,
+///     age: u32,
+///     #[private]
+///     secret: String,
+///     #[collection]
+///     friends: Vec<Person>,
+///     #[storage]
+///     storage: Element,
```

Comment:
so does this mean we're exposing `calimero_storage` to the WASM environment? ..._

Reply:
@danwilliams  it should behave in the same way than in the transactional model, but instead using CRDTs

Reply:
https://github.com/calimero-network/core/pull/764#discussion_r1781443958

Reply:
To confirm, neither the `Element` nor the `Collection` should cause any problems if exposed to the guest, and indeed are meant to be. That does not necessarily mean that we are exposing the `Interface`, but everything else is a grey area. The way in which the guest side works with this storage functionality would be a separate piece of work - this is storage layer only.